File: ferreteria_refactor/backend_api/models/tenant.py
from sqlalchemy import Column, Integer, String, Boolean, DateTime, JSON
from sqlalchemy.dialects.postgresql import JSONB
from ..database.db import Base
from ..utils.time_utils import get_venezuela_now

class Tenant(Base):
    __tablename__ = "tenants"
    __table_args__ = {"schema": "public"}  # Always resides in public schema

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String, nullable=False)
    schema_name = Column(String, unique=True, index=True, nullable=False)
    domain = Column(String, nullable=True, unique=True)
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime, default=get_venezuela_now)
    
    # Subscription / License Logic
    is_demo = Column(Boolean, default=True) # New tenants are Demo by default usually
    license_type = Column(String(20), default="trial", nullable=False, server_default="lifetime")
    trial_days = Column(Integer, default=2, nullable=False, server_default="2")
    trial_ends_at = Column(DateTime, nullable=True)
    subscription_expires_at = Column(DateTime, nullable=True) # NULL means forever or undefined
    license_blocked_reason = Column(String(50), nullable=True)

    # Stores feature flags: {"restaurant": true, "laundry": false}
    config = Column(JSON, default=dict)
    
    # NEW: Specific Sector/Rubro Label
    business_type = Column(String, nullable=True) # e.g. "Abasto", "Ferretería"

    # Module Flags (NEW System)
    has_restaurant_module = Column(Boolean, default=False)
    has_laundry_module = Column(Boolean, default=False)
    has_hardware_module = Column(Boolean, default=False)
    has_services_module = Column(Boolean, default=False)
    has_barbershop_module = Column(Boolean, default=False)
    has_pharmacy_module = Column(Boolean, default=False)

    # Feature flags a la carta — activadas individualmente por tenant desde el panel admin
    # Ejemplo: {"descuento_especial": true, "reporte_avanzado": false}
    feature_flags = Column(JSONB, nullable=False, server_default='{}')

    def __repr__(self):
        return f"<Tenant(name={self.name}, schema={self.schema_name})>"

File: ferreteria_refactor/saas_admin/src/types/tenant.ts
export type LicenseType = 'trial' | 'monthly' | 'annual' | 'lifetime';

export interface Tenant {
    id: number;
    name: string;
    schema_name: string;
    domain: string | null;
    is_active: boolean;
    created_at: string;
    config?: Record<string, unknown>;
    // Subscription Fields
    is_demo: boolean;
    subscription_expires_at: string | null; // ISO Date
    days_remaining?: number; // Calculated helper
    user_count?: number;

    // License Fields
    license_type: LicenseType;
    trial_days: number;
    trial_ends_at: string | null;
    license_blocked_reason: string | null;

    // Module Flags
    has_restaurant_module: boolean;
    has_laundry_module: boolean;
    has_hardware_module: boolean;
    has_services_module: boolean;
    has_barbershop_module: boolean;
    has_pharmacy_module: boolean;

    // Feature flags a la carta
    feature_flags: Record<string, boolean>;
}

export interface FeatureFlagDef {
    label: string;
    description: string;
    category: string;
}

export interface FeatureFlagsRegistry {
    registry: Record<string, FeatureFlagDef>;
    categories: string[];
}

export interface TenantListResponse {
    total: number;
    tenants: Tenant[];
}

export interface CreateTenantDTO {
    name: string;
    schema_name: string;
    domain?: string;
    admin_email: string;
    admin_password: string;
    config?: Record<string, any>;
    // Subscription Fields
    is_demo: boolean;
    subscription_expires_at: string | null;
}

export interface UpdateTenantDTO {
    name?: string;
    domain?: string;
    is_active?: boolean;
    config?: Record<string, unknown>;
    is_demo?: boolean;
    subscription_expires_at?: string | null;

    // License Fields
    license_type?: LicenseType;
    trial_days?: number;
    trial_ends_at?: string | null;
    license_blocked_reason?: string | null;

    // Module Flags
    has_restaurant_module?: boolean;
    has_laundry_module?: boolean;
    has_hardware_module?: boolean;
    has_services_module?: boolean;
    has_barbershop_module?: boolean;
    has_pharmacy_module?: boolean;
}
